lib/browser/source.js

```diff
@@ -11,7 +11,102 @@ class FileSource {
   close() {}
 }
 
-export function getSource(input) {
+class StreamSource {
+  constructor(reader, chunkSize) {
+    // Ensure that chunkSize is an integer and not something else or Infinity.
+    chunkSize = +chunkSize;
+    if (!isFinite(chunkSize)) {
+      throw new Error("cannot create source for stream without a finite value for the `chunkSize` option");
+    }
+    this._chunkSize = chunkSize;
+
+    this._buffer = undefined;
+    this._bufferOffset = 0;
+    this._reader = reader;
+    this._done = false;
+  }
+
+  slice(start, end, callback) {
+    if (start < this._bufferOffset) {
+      throw new Error("requested data is before the reader's current offset");
+    }
+
+    return this._readUntilEnoughDataOrDone(start, end, callback);
+  }
+
+  _readUntilEnoughDataOrDone(start, end, callback) {
+    const hasEnoughData = end <= this._bufferOffset + len(this._buffer);
+    if (this._done || hasEnoughData) {
+      callback(this._getDataFromBuffer(start, end));
+      return;
+    }
+    this._reader.read().then(({ value, done }) => {
+      if (done) {
+        this._done = true;
+      } else if (this._buffer === undefined){
+        this._buffer = value;
+      } else {
+        this._buffer = concat(this._buffer, value);
+      }
+
+      this._readUntilEnoughDataOrDone(start, end, callback);
+    }).catch(err => {
+      throw new Error("Error during read: "+err);
```

Comment:
You should not use `throw` in there. Since this function may be invoked asynchronously, the error would not be catchable. Instead, the error should be passed up to the callback where it is then handled by `Upload#_emitError`. This also applies to the error in `StreamSource#slice`.

Reply:
Fixed this up and also made my callback argument style match the other callbacks. Let me know how it looks!